src/subsystem/drive_controller.cpp

```diff
@@ -0,0 +1,75 @@
+#include "drive_controller.hpp"
+#include <stdio.h>
+using namespace std;
+
+// Separate the implementation from drive_controller.hpp and place here
+
+void DriveController::halt() {
+    accelerate_to(0, 0);
+}
+
+void DriveController::set_motor_acc(char direction, float acc) { 
+    if (direction == 'L') {
+        left_motor_1 = acc;
+        left_motor_2 = acc;
+        left_motor_3 = acc;
+    }
+    if (direction == 'R') {
+        right_motor_1 = acc;
+        right_motor_2 = acc;
+        right_motor_3 = acc;
+    }
+}
+
+// Called whenever current_angle and target_velocity changes
+void DriveController::update_target_velocity() { 
+    target_velocity_y = cos(current_angle * M_PI / 180) * target_velocity_mps;
+    target_velocity_x = sin(current_angle * M_PI / 180) * target_velocity_mps;
+}
+
+float DriveController::get_velocity() { 
+    return sqrt(pow(current_velocity_x, 2) + pow(current_velocity_y, 2));
+}
+
+// moves to target velocities? has to find angle 
+void DriveController::accelerate_to(float velocity_x, float velocity_y) {     
+    float acceleration_x = velocity_x - current_velocity_x;
+    float acceleration_y = velocity_y - current_velocity_y;
+    float target_acceleration = sqrt(pow(acceleration_x, 2) + pow(acceleration_y, 2));
+    float angle = atan(acceleration_y / acceleration_x) * (180 / M_PI);
+    
+    if (acceleration_x != 0 || acceleration_y != 0) {
+        float right_scaling = (90 + angle)/180;
+        set_motor_acc('L', (1 - right_scaling) * target_acceleration);
+        set_motor_acc('R', right_scaling * target_acceleration);
+        
+        current_velocity_x += acceleration_x;
+        current_velocity_y += acceleration_y;
+    }
+    else { 
+        set_motor_acc('L', 0);
+        set_motor_acc('R', 0);
+    }
+}
```

Comment:
I would rework this to not rely on any trigonometry and simply split the system between left and right. The current calculations seem not to work as there's a divide by zero on line 39 when the angle is 0.

Your set functions are fine, but I'd change how the inputs are translated to speed. I'd have two other saved values for `target_left_speed`, `target_right_speed`, `left_speed`, and `right_speed`. In the `update_target_velocity` function, start by setting the target speeds equal to the forward speed. Then, check if there is any turning. If angle is less than 0, multiply the left speed by `abs(angle) / 90`, and do the same for right if the angle is greater than 0.

Reply:
I removed accelerate_to and just put everything in update_motor_acceleration and added the four variables as member variables.  

For update_target_velocity, when angle was less than 0, I multiplied by 90 + left_speed (-90 means 0 speed) and something similar for when angle was > 0.